PP-662, PP-663: Allow an admin to modify the start/end times of a res…

prakashcv13 opened the pull request:
…ervation.

<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Issue-ID
<!--- Replace XXXX below (2 places) with the actual JIRA id -->
* **[PP-662](https://pbspro.atlassian.net/browse/PP-662)**
* **[PP-663](https://pbspro.atlassian.net/browse/PP-663)**
* **[PP-906](https://pbspro.atlassian.net/browse/PP-906)**
* **[PP-703](https://pbspro.atlassian.net/browse/PP-703)**

#### Problem description
* Allow changing of start and end times of a reservation.
* The requirements are listed in the [UCR](https://pbspro.atlassian.net/wiki/pages/viewpage.action?pageId=51810081&src=jira)

#### Cause / Analysis
* Until now reservation times cannot be changed.

#### Solution description
* A new CLI pbs_ralter is added as per the [External Interface Design](https://pbspro.atlassian.net/wiki/pages/viewpage.action?pageId=51810117&src=jira)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -1005,3 +1005,17 @@ def test_check_alternate_nodes_standing_resv(self):
+
+    def test_conflict_standing_resv_instances(self):
+        """
+        """
```
Add a docstring to this test.

bhroam commented:
Windows CI failed due to an unresolved symbol:
         dis_read.obj : error LNK2019: unresolved external symbol _decode_DIS_ModifyResv referenced in function _dis_request_read
         ..\..\..\win_build\src\server\Release\pbs_server.exe : fatal error LNK1120: 1 unresolved externals

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -548,24 +573,33 @@ req_confirmresv(struct batch_request *preq)
-	 * and the reservation to the associated vnodes
+	if (is_being_altered & END_TIME_MODIFIED) {
+		if (gen_task_EndResvWindow(presv)) {
+			req_reject(PBSE_SYSTEM, 0, preq);
```
it is being removed in gen_future_deleteResv().

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -548,24 +573,33 @@ req_confirmresv(struct batch_request *preq)
 	 */
-	if (!is_degraded && (rc = gen_task_Time4resv(presv)) != 0) {
+	if (!is_degraded && (is_being_altered != END_TIME_MODIFIED) && 
+		(rc = gen_task_Time4resv(presv)) != 0) {
```
it is being removed in gen_task_Time4resv().

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -596,12 +629,12 @@ req_confirmresv(struct batch_request *preq)
+			sprintf(buf, "%.240s CONFIRMED",  presv->ri_qs.ri_resvID);
 		}
 
 		if (strlen(log_buffer) > 0) {
```
done!

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -611,18 +644,104 @@ req_confirmresv(struct batch_request *preq)
-	if (presv->ri_wattr[RESV_ATR_interactive].at_flags & ATR_VFLAG_SET) {
-		(void)sprintf(buf, "requestor=%s@%s Interactive=%ld",
+	if (presv->ri_wattr[RESV_ATR_interactive].at_flags & ATR_VFLAG_SET)
+		presv->ri_wattr[RESV_ATR_interactive].at_flags &= ~ATR_VFLAG_SET;
```
right. removed the if.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -611,18 +644,104 @@ req_confirmresv(struct batch_request *preq)
+	if (is_being_altered) {
+		/*
+		 * If the reservation's start time is being altered after the reservation had started
+		 * running, we need to stop the reservation queue as it would been started at the
```
did the rephrase. hope it makes sense now.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -611,18 +644,104 @@ req_confirmresv(struct batch_request *preq)
+		 * reservation's start time is changed - from running.
+		 * The reservation went to CO from RN while being altered, that means the reservation
+		 * had resources assigned. We should decrement their usages until it starts running
+		 * again, where the resources will be accounted again.
```
this is not a race condition. Maybe I can make the comment more clear. 
If the reservation was running and its start time was changed, there is a possibility that the reservation will start late, so we have to take back the resources.
And we stop the queue so that the jobs in it don't start running.

prakashcv13 commented on src/server/req_rescq.c:
```diff
@@ -611,18 +644,104 @@ req_confirmresv(struct batch_request *preq)
+	eval_resvState(presv, RESVSTATE_alter_failed, 0, &state, &sub);
+	/*
+	 * While requesting alter, substate was retained, so we use the same here.
+	 */
```
done.

prakashcv13 commented on src/server/svr_jobfunc.c:
```diff
@@ -3392,8 +3399,17 @@ Time4occurrenceFinish(resc_resv *presv)
+
+	/* If the start time of the reservation was altered, copying from RESV_ATR_start
+ 	 * will make the next instance to have it's start time altered so take the start
+ 	 * time from the ri_alter_stime. */
```
just copy. corrected.

prakashcv13 commented on src/server/svr_jobfunc.c:
```diff
@@ -3801,9 +3832,22 @@ eval_resvState(resc_resv *presv, enum resvState_discrim s, int relVal,
 	*pstate = presv->ri_qs.ri_state;
 	*psub = presv->ri_qs.ri_substate;
 
+        sprintf(log_buffer, "", presv->ri_qs.ri_stime, presv->ri_qs.ri_etime, time_now);
```
this was a debug printf(). Removed it.

prakashcv13 commented on src/server/svr_jobfunc.c:
```diff
@@ -3801,9 +3832,22 @@ eval_resvState(resc_resv *presv, enum resvState_discrim s, int relVal,
 		if (relVal == 0) {
-			if (presv->ri_qs.ri_etime > time_now) {
+			if (*pstate == RESV_BEING_ALTERED) {
+				/* Altering a reservation before its start time */
```
changed the comment.

prakashcv13 commented on src/server/svr_jobfunc.c:
```diff
@@ -3801,9 +3832,22 @@ eval_resvState(resc_resv *presv, enum resvState_discrim s, int relVal,
+					*psub = RESV_CONFIRMED;
+				}
+				/* Altering a reservation after its start time but before end time */
+				else if (presv->ri_qs.ri_etime >= time_now) {
```
not sure why it was there. Removed it.

prakashcv13 commented on src/server/svr_jobfunc.c:
```diff
@@ -3850,6 +3894,9 @@ eval_resvState(resc_resv *presv, enum resvState_discrim s, int relVal,
 		*pstate = RESV_UNCONFIRMED;
 		*psub = RESV_UNCONFIRMED;
 	}
+	else if (s == RESVSTATE_alter_failed)
```
done.

prakashcv13 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -7697,6 +7702,75 @@ def pbs_api_as(self, cmd=None, obj=None, user=None, **kwargs):
+        if runas is not None:
+            prefix += ' as ' + str(runas)
+        prefix += ': '
+        prefix += resvid
```
combined.

prakashcv13 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -7697,6 +7702,75 @@ def pbs_api_as(self, cmd=None, obj=None, user=None, **kwargs):
+            if c < 0:
+                return c
+            a = self.utils.convert_to_attrl(attrib)
+            rc = pbs_modify_resv(c, resvid, a, extend)
```
I have just changed the mock file.

prakashcv13 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12793,6 +12867,8 @@ def select(self):
         elif 'select' in self.attributes:
             self._select = PbsTypeSelect(self.attributes['select'])
+        elif 'resv_nodes' in self.attributes:
+            self._select = self.attributes['resv_nodes']
```
right, but I needed the full string for comparison, so I did this.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+import time
```
removed the import.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+
+        param ExpectSuccess: Whether the caller expects the reservation
+                             to be successful or not.
+        type  ExpectSuccess: int.
```
it is not used as a boolean. I have changed the variable name and provided a description of this argument.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        start = int(time.time()) + offset
+        end = start + duration
+
+        if (standing):
```
removed all the () from all if statements.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            msg += " recurrence_rrule=" + rrule + " timezone="
+            msg += self.tzone
+
+        self.server.log_match(msg, regexp=True, interval=2, max_attempts=30)
```
some of the log_match need the regexp, I have retained those and removed the unwanted ones.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        else:
+            msg = ".*Resv;" + rid + ";Reservation denied."
+            self.server.log_match(msg, regexp=True, interval=2,
+                                  max_attempts=30)
```
we need to escape rrule and tzone as they have special characters.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+                self.logger.info("Sleeping.")
+
+        attrs = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')}
+        self.server.expect(RESV, attrs, id=rid)
```
expect() creates a lots of logs, so I prefer sleep here.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        if (standing):
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        else:
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_FINISHED|6')}
```
I saw that I am using this function only for standing reservations, so the code has changed and it doesn't have the race condition.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        else:
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_FINISHED|6')}
+        self.server.expect(RESV, attrs, id=rid)
```
the above comment stands here as well.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        type  end: int.
+        """
+        next_start = start + 3600
+        next_end = end + 3600
```
added a comment.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        type  start: int.
+
+        param end: End time of the reservation.
+        type  end: int.
```
to avoid a call to stat. :)

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+                       Default - No.
+        type  alter_s: bool.
+
+        param alter_e: Whether the caller intents to change the end time.
```
right. corrected.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+
+        param ExpectSuccess: Whether the caller expects the reservation
+                             to be successful or not.
+                             Default: Yes.
```
change the variable name. provided an explanation as well.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        type  interactive: int.
+
+        param sequence: To check the number of logs corresponding to alter.
+                        Default: 1
```
changed.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            attrs.update({'reserve_end': new_end_conv})
+
+        if (interactive > 0):
+            attrs.update({'interactive': interactive})
```
depending on which boolean is set, the values get added to the dictionary. I find this intuitive.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+                elif (ExpectSuccess == 2):
+                    msg = "pbs_ralter: " + rid + " UNCONFIRMED"
+                else:
+                    msg = "pbs_ralter: " + rid + " DENIED"
```
changed the variable name and provided an explanation.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            else:
+                msg = "pbs_ralter: " + rid + " ALTER REQUESTED"
+
+            ret = self.server.alterresv(rid, attrs)
```
would you like me to create a PTL ticket for changing alterjob as for my purposes I want the full dictionary being returned from alterresv.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+                if (rid[0] == 'S'):
+                    standing = 1
+
+                if (standing):
```
combined.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        else:
+            try:
+                self.server.alterresv(rid, attrs)
+            except PbsResvAlterError, e:
```
changed.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            else:
+                self.logger.error("Reservation alter allowed when it should" +
+                                  "not be.")
+                self.assertFalse(True)
```
done.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+
+        1. Move back an advance reservation's start time (empty reservation)
+        2. Move forward an advance reservation's start time (empty reservation)
+        3. Move forward an advance reservation's start time (with a job in it)
```
didn't know they will be confusing to you. Changed the comments.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        that has not started running.
+
+        1. Move back an advance reservation's end time (empty reservation)
+        2. Move back an advance reservation's end time (empty reservation)
```
yes.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        self.check_resv_running(rid, offset)
+
+        # Submit a job to the reservation.
+        jid = self.submit_job_to_resv(rid)
```
I am using it in some test cases. In others I have removed it.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+
+        # Wait for the reservation instance to finish.
+        new_duration = new_end - new_start
+        self.check_resv_finish(rid, new_duration, standing=True)
```
right. made the changes.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        self.check_resv_running(rid, offset)
+
+        new_start, new_end = self.alter_a_reservation(rid, start, end, shift,
+                                                      alter_e=True, confirm=0)
```
changed.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+                                 ExpectSuccess=3)
+        self.alter_a_reservation(rid1, start1, end1, shift, alter_e=True,
+                                 ExpectSuccess=3, interactive=5, sequence=2)
+
```
extended the test case to test that as well.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+
+        attrs = {'resources_available.ncpus': (GT, 0)}
+        vnode_dict = self.server.filter(NODE, attrs)
+        vnode_list = vnode_dict.values()[0]
```
yeah, but I like the natural vnode to have 0 cpus. So, I have not made any changes here.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+        attrs = {'resv_nodes': (MATCH_RE, re.escape(free_node))}
+        self.server.expect(RESV, attrs, id=rid1)
+
+    def test_check_alternate_nodes_standing_resv(self):
```
added two new test cases.

prakashcv13 commented:
I am not sure why pep8 is failing in travis. I had confirmed the changes to be pep8 compliant on ubuntu and centos7 systems before committing. Could someone please let me know if I am missing anything.

Centos 7
=======
[root@d_server tests]# cat /etc/redhat-release 
CentOS Linux release 7.3.1611 (Core) 
[root@d_server tests]# pep8 functional/pbs_ralter.py
[root@d_server tests]# pep8 ../fw/ptl/lib/pbs_testlib.py 
[root@d_server tests]# 

ubuntu 16.04
==========
prakv@blrlap551:~/github/pbspro$ cat /etc/lsb-release 
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.2 LTS"
prakv@blrlap551:~/github/pbspro$ pep8 test/tests/functional/pbs_ralter.py
prakv@blrlap551:~/github/pbspro$ pep8 test/fw/ptl/lib/pbs_testlib.py 
prakv@blrlap551:~/oldhome/work/github/prakashcv13/master/pbspro$ 

Thanks,
Prakash

arungrover commented:
@prakashcv13 it's the checksign that is failing not pep8. Please sign the commit

prakashcv13 commented:
Thank you @arungrover, I had resolved conflicts on github and merged the changes that had created an unsigned commit. Now it looks ok.

bhroam commented:
@prakashcv13 I'm still concerned about the discrepancy between alterjob() and alterresv().  I think alterjob() does it right.  You are returning the value of run_cmd().  That only happens in CLI mode.  In API mode, you return ret but I don't think ret is set to anything.  There really isn't the same dictionary to return in API mode.  You have to return a value that is consistent between both.  Due to this, just returning the rc is probably best.

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -968,15 +994,54 @@ def test_check_alternate_nodes_advance_resv(self):
+
+        attrs = {'resources_available.ncpus': (GT, 0)}
+        vnode_dict = self.server.filter(NODE, attrs)
+        vnode_list = vnode_dict.values()[0]
```
I understand that you like having the natural vnode with 0 cpus, but if you used it you wouldn't have to do this.  This just adds to confusion to the test and takes extra test time.

prakashcv13 replied:
Ok, removed :)

prakashcv13 commented on src/cmds/pbs_ralter.c:
```diff
@@ -26,15 +49,17 @@ static struct attrl *attrib = NULL;
+ * @param[in]  argc  - the number of arguments to be parsed.
+ * @param[in]  argv  - the argument array.
+ * @param[out] attrp - attribute list sent to the caller.
  * @param[out] attrp - attribute list sent to the caller.
```
will remove.

prakashcv13 commented on src/cmds/pbs_ralter.c:
```diff
@@ -26,15 +49,17 @@ static struct attrl *attrib = NULL;
+ * @param[in]  argv  - the argument array.
+ * @param[out] attrp - attribute list sent to the caller.
  * @param[out] attrp - attribute list sent to the caller.
+ * @param[out] dest  - server option
```
again a copy paste from pbs_rsub.c. Will clarify.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        else:
+            attrs = {'reserve_state': (MATCH_RE, 'RESV_FINISHED|6')}
+        self.server.expect(RESV, attrs, id=rid)
```
the occurrences are an hour apart, however, I like the idea of using the offset, so I will use it here.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1007 @@
+            attrs.update({'reserve_end': new_end_conv})
+
+        if (interactive > 0):
+            attrs.update({'interactive': interactive})
```
I will be using the assignment instead of update.

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -1942,7 +2000,7 @@ find_correct_nodes(status *policy, server_info *sinfo, queue_info *qinfo, resour
 	 * to the job.  We won't need to search for them
```
it was being used somewhere else. removed it now.

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -2022,13 +2080,48 @@ void get_resresv_spec(resource_resv *resresv, selspec **spec, place **pl)
 
+/**
+ * @brief
+ * mark_resv_nodes_for_alter 	- this function marks nodes for consideration in
```
done!

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -2022,13 +2080,48 @@ void get_resresv_spec(resource_resv *resresv, selspec **spec, place **pl)
+ * @return void
+ */
+void
+mark_resv_nodes_for_alter(node_info **unassoc_nodes, node_info **resv_nodes, int flag)
```
I have changed the whole function and gave it a new meaning.

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -2022,13 +2080,48 @@ void get_resresv_spec(resource_resv *resresv, selspec **spec, place **pl)
+{
+	int		i = 0;
+	node_info	*temp = NULL;
+
```
added the checks.

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -2022,13 +2080,48 @@ void get_resresv_spec(resource_resv *resresv, selspec **spec, place **pl)
+	node_info	*temp = NULL;
+
+	for (i = 0; unassoc_nodes[i] != NULL; i++) {
+		if (flag == PASS_ONE) {
```
corrected.

prakashcv13 commented on src/scheduler/check.c:
```diff
@@ -2022,13 +2080,48 @@ void get_resresv_spec(resource_resv *resresv, selspec **spec, place **pl)
+	for (i = 0; unassoc_nodes[i] != NULL; i++) {
+		if (flag == PASS_ONE) {
+			temp = find_node_info(resv_nodes, (*unassoc_nodes)->name);
+			if (temp)
```
ok. done!

prakashcv13 commented on src/scheduler/node_info.c:
```diff
@@ -2018,38 +2018,38 @@ find_nspec_by_rank(nspec **nspec_arr, unsigned int rank)
- * @retval	1	: if the nodespec can be satisfied
- * @retval	0	: if not
+ * @retval	1	  : 	if the nodespec can be satisfied
+ * @retval	0	  : 	if not
```
the ':' s are there in the whole file, so keeping them.

prakashcv13 commented on src/scheduler/node_info.c:
```diff
@@ -2105,8 +2105,18 @@ eval_selspec(status *policy, selspec *spec, place *placespec,
-			*nspec_arr = NULL;
+			if (resresv->is_resv && resresv->resv && resresv->resv->check_alternate_nodes) {
+				mark_resv_nodes_for_alter(ninfo_arr, resresv->ninfo_arr, PASS_TWO);
+				rc = eval_placement(policy, spec, ninfo_arr, pl, resresv, pass_flags, nspec_arr, err);
```
I have changed the way we go about this now and tried to be efficient. What I will be doing now is pass all the nodes sorted with nodes belonging to the reservation ahead of the other nodes. That way we make only one pass through the nodes.

prakashcv13 commented on src/server/req_modify.c:
```diff
@@ -760,13 +769,23 @@ req_modifyReservation(struct batch_request *preq)
+						(temp != presv->ri_wattr[RESV_ATR_start].at_val.at_long)) {
+						if (!is_standing || (temp < next_occr_start &&
+							((temp + presv->ri_wattr[RESV_ATR_duration].at_val.at_long) < next_occr_start))) {
+							send_to_scheduler = RESV_START_TIME_MODIFIED;
```
right, removed the unnecessary check.

prakashcv13 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12868,7 +12867,7 @@ def select(self):
             self._select = PbsTypeSelect(self.attributes['select'])
         elif 'resv_nodes' in self.attributes:
-            self._select = self.attributes['resv_nodes']
+            return self.attributes['resv_nodes']
```
removed this and added another function which returns the reservation nodes.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -188,28 +191,29 @@ def check_resv_finish(self, rid, duration, standing=False):
-            attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
-        else:
-            attrs = {'reserve_state': (MATCH_RE, 'RESV_FINISHED|6')}
+        attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
```
changed.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -262,18 +266,21 @@ def alter_a_reservation(self, rid, start, end, shift,
+                             =0 - PbsResvAlterError exception will be raised,
+                                  so check for appropriate error response.
+                            !=0 - No exception, check for appropriate CLI
+                                  response.
```
listed a description of the messages.

prakashcv13 commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -391,42 +395,47 @@ def alter_a_reservation(self, rid, start, end, shift,
+                elif e.rc == self.PBSE_UNKRESVID:
                     msg = "pbs_ralter: Unknown Reservation Id"
-                elif (e.rc == self.PBSE_INCORRECT_USAGE):
+                elif e.rc == self.PBSE_STDG_RESV_OCCR_CONFLICT:
```
I am using only CLI mode in this test, so it should be fine.

prakashcv13 commented:
Hi @bhroam, 
Thank you again for a thorough code review. With regards to making alterjob and alterresv same, I do see what you are saying is correct and have made the changes accordingly.

Thanks!

bhroam commented on src/scheduler/check.h:
```diff
@@ -226,33 +226,18 @@ schd_resource *zero_res(void);
  *	returns void
  */
-void mark_resv_nodes_for_alter(node_info **unassoc_nodes, node_info **resv_nodes, int flag);
+void reorder_nodes_set(node_info **superset, node_info **subset);
```
I wouldn't call this nodes_set.  There is something called 'node_set' in the scheduler currently.  Maybe move_subset_to_front() or something like that.

prakashcv13 replied:
This function will be removed as we discussed offline.

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -5619,3 +5611,34 @@ check_node_array_eligibility(node_info **ninfo_arr, resource_resv *resresv, plac
+		return;
+
+	for (i = 0; superset[i] != NULL; i++) {
+		temp = find_node_info(subset, superset[i]->name);
```
use find_node_by_rank().  Avoid searching by strings as much as possible.

prakashcv13 replied:
done!

bhroam commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -325,21 +322,21 @@ def alter_a_reservation(self, rid, start, end, shift,
-            ret = self.server.alterresv(rid, attrs)
-            self.assertEqual(ret['rc'], 0)
+            rc = self.server.alterresv(r, attrs)
+            self.assertEqual(rc, 0)
```
Do you need check for this?  Won't alterresv() raise an exception if something bad happened?

prakashcv13 replied:
Removed this check.

bhroam commented:
@prakashcv13 you might take a look at what codacy is reporting.  There are a couple things there that would be nice to get fixed.

prakashcv13 commented on src/scheduler/node_info.c:
```diff
@@ -2867,33 +2853,34 @@ eval_simple_selspec(status *policy, chunk *chk, node_info **pninfo_arr,
+	nspec		**nsa = NULL;		/* the nspec array to hold node solution */
+	resource_req	*specreq_noncons = NULL;/* non-consumable resources requested by spec */
+	resource_req	*specreq_cons = NULL;	/* consumable resources requested by spec */
+	resource_req	*req, *prevreq = NULL;	/* used to determine if we're done */
```
done!

prakashcv13 commented on src/scheduler/node_info.c:
```diff
@@ -2979,9 +2966,14 @@ eval_simple_selspec(status *policy, chunk *chk, node_info **pninfo_arr,
 
+	if (resresv->is_resv && resresv->resv && resresv->resv->check_alternate_nodes) {
+		reorder_nodes_set(ninfo_arr, resresv->ninfo_arr);
+		resresv->resv->check_alternate_nodes = 0;
```
We discussed and decided offline, we will be merging the new function reorder_nodes_set() to reorder_nodes() and call it in eval_placement().

prakashcv13 commented:
@bhroam, 

I have addressed all your comments and made changes as per our offline discussion. codacy is also reporting Green now.

Thanks,
Prakash

prakashcv13 commented:
@bhroam,

The recent commit is for fixing the issues that @varunsonkar had found in his offline testing. 
1. pbs_ralter crashes when trying to alter a degraded reservation and the request is denied by th scheduler.
2. requested end time was not getting compared with the current end time like we do for the requested start time.

Thanks,
Prakash

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -2244,24 +2244,24 @@ eval_placement(status *policy, selspec *spec, node_info **ninfo_arr, place *pl,
+	int			k = 0;
+	int			tot = 0;
+	int			c = -1;
+	int			cur_flt_lic = resresv->server->flt_lic;
```
This will crash if resresv or resresv->server is NULL.  Assign this after you've already checked reseresv for NULL.

prakashcv13 replied:
done!

bhroam commented on src/scheduler/node_info.c:
```diff
@@ -4623,6 +4620,22 @@ reorder_nodes(node_info **nodes, resource_resv *resresv)
 	node_array[0] = NULL;
 	nptr = node_array;
 
+	if (resresv->is_resv && resresv->resv && resresv->resv->check_alternate_nodes) {
```
Check pointers against NULL, so if (ptr != NULL) not if (ptr)

prakashcv13 replied:
done!

bhroam commented on src/server/req_modify.c:
```diff
@@ -704,8 +703,7 @@ req_modifyReservation(struct batch_request *preq)
 	extern char	*msg_stdg_resv_occr_conflict;
-
-	presv = chk_rescResv_request(preq->rq_ind.rq_modify.rq_objname, preq);
+	resc_resv	*presv = chk_rescResv_request(preq->rq_ind.rq_modify.rq_objname, preq);
```
This will crash if preq is NULL.

prakashcv13 replied:
done!

bhroam commented on src/server/req_modify.c:
```diff
@@ -903,13 +901,11 @@ modify_resv_attr(resc_resv *presv, svrattrl *plist, int perm, int *bad)
 	int	   modified = 0;
 	attribute  newattr[(int)RESV_ATR_LAST] = {0};
-	attribute *pattr = NULL;
+	attribute *pattr = presv->ri_wattr;
```
This will crash if presv is NULL.

bhroam commented on src/server/req_rescq.c:
```diff
@@ -338,7 +338,7 @@ req_confirmresv(struct batch_request *preq)
 	 */
 	if (strcmp(preq->rq_extend, PBS_RESV_CONFIRM_FAIL) == 0) {
-		if (is_degraded) {
+		if (is_degraded && !is_being_altered) {
```
What happens if you alter a degraded reservation?

prakashcv13 replied:
This change is just for that case (altering a degraded reservation). It should be treated that the alter request has failed.

bhroam replied:
OK I walked through the logic.  It looks good.

bhroam commented:
The code looks good now.  I sign off.  Codacy is complaining, but for the life of me I can't figure out why.  Neither check is wrong.

dilip-krishnan commented on src/cmds/pbs_ralter.c:
```diff
@@ -0,0 +1,272 @@
+	fprintf(stderr, usage);
+	fprintf(stderr, usag2);
+}
+
```
please add doxygen.

dilip-krishnan commented:
Hi @prakashcv13 ,
   Apart from the changes requested during sit down review, i don't have any further comments. Rest of the code looks good to me.

Warm regards

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        else:
+            self.logger.info('Timezone not set, using Asia/Kolkata')
+            self.tzone = 'Asia/Kolkata'
+
```
In case the "PBS_TZID" is not set in conf file or in the environment. Set it to system default timezone.

prakashcv13 replied:
I believe you meant, setting the system timezone. As it wouldn't affect the outcome of the tests, I will leave this as it is.

varunsonkar replied:
If you say that it does not have any affect then I would suggest to remove it.
It does not look right to set to 'Asia/Kolkata'.
self.tzone = 'Asia/Kolkata'

prakashcv13 replied:
We need to set this to a valid timezone for standing reservations to work. I am setting it to Asia/Kolkata.
Could you please clarify what does not look right in this?

varunsonkar replied:
Ok fine I got it. It is ok.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        rc = self.server.manager(MGR_CMD_SET, SERVER, {
+            'log_events': 4095}, expect=True)
+        self.assertTrue(rc)
+
```
self.assertTrue(rc) is not required if you are doing "expect=True". Any exception will be caught in case qmgr set command fails.

prakashcv13 replied:
done!

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        It also checks for the corresponding server and accounting logs.
+
+        param offset: time in seconds after which the reservation should start.
+        type  offset: int
```
Suggest you to get rid off offset. Make the checks iterative.
Any check in PTL for state change is iterative.

prakashcv13 replied:
I don't think you have understood the purpose of offset in this function. It is the time in seconds after which the reservation should start. It is already mentioned in the comment.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+
+        param ExpectSuccess: Whether the caller expects the submission to be
+                             successful or not.
+        type  ExpectSuccess: int.
```
use ":param" and ":type". colon is missing.

prakashcv13 replied:
done.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        type  offset: int.
+
+        param display: Whether to display a message or not, default - Yes.
+        type  display: bool
```
Same comments as above for "param", "type" and "offset"

prakashcv13 replied:
done.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+                self.logger.info("Sleeping.")
+
+        attrs = {'reserve_state': (MATCH_RE, 'RESV_RUNNING|5')}
+        self.server.expect(RESV, attrs, id=rid, offset=offset)
```
expect is iterative. Iterations can be overwritten. Fixed waiting is not desirable.

prakashcv13 replied:
If that is the case, why does the expect method provides an offset option. I will keep the expect as it is.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        if duration > 0:
+            self.logger.info("Waiting for reservation to finish.")
+        attrs = {'reserve_state': (MATCH_RE, 'RESV_CONFIRMED|2')}
+        self.server.expect(RESV, attrs, id=rid, offset=duration-5, interval=2)
```
Do not use offset. Use iteration.

prakashcv13 replied:
Same as above.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        type  start: int.
+
+        param end: End time of the first occurrence of the reservation.
+        type  end: int.
```
same comments for param and type

prakashcv13 replied:
done.

varunsonkar commented on test/tests/functional/pbs_ralter.py:
```diff
@@ -0,0 +1,1112 @@
+        """
+        offset = 120
+        duration = 120
+        shift = 60
```
These values are way too high for automation purposes.
The test suite will consume a lot of time to complete. Please optimize them to the best.
This comment applies for all the test cases. Please take a look and optimize the time whatever best can be done.

prakashcv13 replied:
Hmmm. This can be done. I have reduced the duration and offsets to reduce the run time of the test suits.

prakashcv13 commented:
Thank you @dilip-krishnan and @varunsonkar for your inputs -
@dilip-krishnan,
Below is the summary of our sit-down review - 
previously existing multi-line comments in reservation.h - to be changed as per our coding standards.
doxygen for  decode_DIS_ModifyResv(), to be just above the function definition
check_alternate_nodes bit-field to have an explanation in data_types.h
doxygen for handle_attribute_errors().

Hi @varunsonkar,
I have addressed to some of your comments and for others, I have replied as per my understanding.

Thanks,
Prakash

prakashcv13 commented:
Hi @bhroam,

I am closing the multi-line comments on the RHS in reservation.h now.

Thanks,
Prakash

bhroam commented:
reservation.h looks good.  There are two instances in data_types.h that needs to be fixed as well.

prakashcv13 commented:
@bhroam, corrected data_types.h as well.

dilip-krishnan commented:
Hi @prakashcv13 ,
   Changes look good.

prakashcv13 commented:
Hi @bhroam,

I have all the required approvals for this PR, have squashed my commits and rebased the branch. Could you please go ahead and merge it.

Thanks,
Prakash

bhroam commented:
All merged!

prakashcv13 commented:
Thank you @bhroam.
